Add propd to abundance_differential_filter subworkflow

suzannejin opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

pinin4fjords commented on subworkflows/nf-core/abundance_differential_filter/main.nf:
```diff
@@ -56,45 +57,80 @@ workflow ABUNDANCE_DIFFERENTIAL_FILTER {
 
-    // Perform normalization and differential analysis
+    // ----------------------------------------------------
+    // Perform differential analysis with limma
```
```suggestion
    // Run Limma just once to generate a normalised matrix, configure so 
    // that we don't pass subsetting arguments anddisregard differential 
    // outputs 
```

pinin4fjords commented on subworkflows/nf-core/abundance_differential_filter/main.nf:
```diff
@@ -56,45 +57,80 @@ workflow ABUNDANCE_DIFFERENTIAL_FILTER {
+    )
+
+    // ----------------------------------------------------
+    // Perform differential analysis with deseq2
```
```suggestion
    // Run DESeq2 just once to generate a normalised matrix, configure so 
    // that we don't pass subsetting arguments and disregard differential 
    // outputs 
```

pinin4fjords replied:
The normalisations don't get passed to the differential modules, so they should sit separately (which is why they were positioned distinctly before). I'm fine with your arrangement, but if we're mixing them we should probably label them appropriately actually explain why it's done this way.

suzannejin replied:
Regarding this, are the normalized matrices in the NORM modules the exact matrices that are used in the DE modules? If so, maybe it would be nice at some point to rearrange the module so NORM -> goes to DE?

pinin4fjords replied:
So, DESeq2 and Limma Voom output normalised matrices (they take raw values), so they don't need to be passed normalised matrices.

So, originally on differentialabundance we just took the first output of DESEQ2_DIFFERENTIAL and used the normalised matrix from there. But it caused problems, since that first contrast might have options (the exclude samples ones) that lead to the matrix being subsetted, so the normalised matrix we got did not always contain all samples.

For that reason we developed this workaround, where the differential module is run via a separate alias, just once, not passing any subsetting options (see the config here: https://github.com/nf-core/differentialabundance/blob/3dd360fed0dca1780db1bdf5dce85e5258fa2253/conf/modules.config#L162), and we take the normalised output from there, disregarding the differential.

suzannejin replied:
Why for DIFF you do have cases where you remove samples, whereas for NORM you want the full set? Is it because you need them for GSEA? Or that the plotting need the entire set?

pinin4fjords commented on subworkflows/nf-core/abundance_differential_filter/main.nf:
```diff
@@ -56,45 +57,80 @@ workflow ABUNDANCE_DIFFERENTIAL_FILTER {
+        norm_inputs.contrasts.filter{it[0].method == 'limma'}.first(),
+        norm_inputs.samples_and_matrix.filter{it[0].method == 'limma'}
+    )
+
```
```suggestion
//
// Perform differential analysis with limma
//

```

pinin4fjords commented on subworkflows/nf-core/abundance_differential_filter/main.nf:
```diff
@@ -56,45 +57,80 @@ workflow ABUNDANCE_DIFFERENTIAL_FILTER {
-    ch_normalised_matrix = DESEQ2_NORM.out.normalised_counts.mix(LIMMA_NORM.out.normalised_counts)
-    ch_model = DESEQ2_DIFFERENTIAL.out.model.mix(LIMMA_DIFFERENTIAL.out.model)
+    // ----------------------------------------------------
+    // Recollect results
```
```suggestion
    // Collect results
```

pinin4fjords commented on subworkflows/nf-core/abundance_differential_filter/main.nf:
```diff
@@ -109,8 +145,14 @@ workflow ABUNDANCE_DIFFERENTIAL_FILTER {
     )
 
     emit:
-    results_genewise           = ch_results.map{meta, results -> [meta, results] }
```
good spot

pinin4fjords commented on subworkflows/nf-core/abundance_differential_filter/meta.yml:
```diff
@@ -122,5 +132,8 @@ authors:
   - "@suzannejin"
+  - "@roskamsh"
+  - "@syangaskhan"
+  - "@WackerO"
```
Maybe add yourself as a maintainer?

Did all these authors contribute to your version of the differential subworkflow?

suzannejin replied:
The last three they actually helped during the hackathon to call the multi-channel logic on DESeq2 and limma.
I'm not exactly sure up to which point the contributions should be added as author here in the meta, but wondering that it is anyway nice to credit them

pinin4fjords reviewed (changes requested):
Good work, minor comments

suzannejin commented:
@pinin4fjords Thank you for the review!
I applied your suggestions and added some more comments to make the intentions behind NORM clear.

I also changed PROPR_PROPD to PROPD_DIFFERENTIAL to remain consistent in the naming of DIFFERENTIAL modules inside the subworkflow. But I recall that it is usually not recommended to change modules names. What do you think about this?

pinin4fjords commented:
> @pinin4fjords Thank you for the review! I applied your suggestions and added some more comments to make the intentions behind NORM clear.
> 
> I also changed PROPR_PROPD to PROPD_DIFFERENTIAL to remain consistent in the naming of DIFFERENTIAL modules inside the subworkflow. But I recall that it is usually not recommended to change modules names. What do you think about this?

DESEQ2_DIFFERENTIAL is the actual module name. It was named that way in anticipation of modules for other parts of DESeq2 (though that hasn't happened so far). No need to alias your module unless you're using it in more than one place.
